### PR TITLE
[BREAKING] Disallow legacy constructor

### LINKED ISSUE #4106: Remove workaround for parsing the ``constructor`` keyword.
``constructor`` is currently not a keyword, but parsed as literal. This is a workaround that should be removed for 0.5.0.

Related: #2182 (with PR #3627).

Part of #3656.

### PR DESCRIPTION
**Pull request by erak**
Closes #4106.

### CODE REVIEW

**Comment by axic** [comment]
Sorry I forgot to assign the issue to myself, but did mention the last call I started this work. Anyway it is not a big amount of duplicate effort :)

However we should only merge this after #4208

**Comment by erak** [comment]
@axic Oh sorry, did not get this in the last call. We should try to keep the project board up to date ;) 
So, should we use with this branch / PR?

I agree, let's merge the other PR first.

**Comment by axic** [comment]
@bit-shift I will have a look at the differences between the two PRs. Do you want to review #4208 ?

**Comment by axic** [comment]
My local version of these changes are all included int this PR.

**Comment by erak** [comment]
@christianparpart I removed the expected shadowing warning of ``override_constructor.sol`` and ``constructor_old.sol``. The old version of ``override_constructor.sol`` was basically testing that you can have a function that is called like the constructor of a contracts base contract. Since this construct is not possible anymore, we also should not warn anymore.
And getting the warning if `` contract A { function A() public {} }``  feels a little bit like clutter.

File: libsolidity/analysis/SyntaxChecker.cpp
Comment (chriseth) [inline libsolidity/analysis/SyntaxChecker.cpp]: Should be `FunctionDefinition const* const function`, but I think the last `const` does not really help readability and thus could be dropped.

File: libsolidity/analysis/SyntaxChecker.cpp
Comment (chriseth) [inline libsolidity/analysis/SyntaxChecker.cpp]: Please use the original `v050` message and not the message of the pre-050 warning. Also, the indentation and line breaks are not according to coding style.

File: libsolidity/analysis/SyntaxChecker.cpp
Comment (chriseth) [inline libsolidity/analysis/SyntaxChecker.cpp]: Please don't add `!_function.isConstructor()`. Currently, it looks like this is a good thing to add, but it might hide errors that are introduced when we change the logic of `isConstructor`.

File: libsolidity/parsing/Parser.cpp
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp]: I think the name should still be empty.

File: libsolidity/parsing/Parser.cpp
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp]: Is this to still allow `function constructor(...) { ..}`? I don't think we have to support that - `constructor` is now a real keyword and thus functions cannot be named `constructor`.

File: test/libsolidity/syntaxTests/constructor/overriding_constructor.sol
Comment (chriseth) [inline test/libsolidity/syntaxTests/constructor/overriding_constructor.sol]: Yep, that is exactly what we should test here!

**Review by chriseth — CHANGES_REQUESTED** [review]
Is there a test for
```solidity
contract C {
function constructor() { ... }
}
```
?

File: libsolidity/parsing/Parser.cpp
Comment (erak) [reply libsolidity/parsing/Parser.cpp]: Yes, the idea was to still allow it, but warn about it's usage. But you're right, since this is keyword now, we should not allow it anymore.

Comment (chriseth) [reply libsolidity/parsing/Parser.cpp]: `!result.isConstructor` can be removed (it is an `else if`).

File: test/libsolidity/SolidityEndToEndTest.cpp
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp]: Any reason for this change?

File: test/libsolidity/SolidityParser.cpp
Comment (chriseth) [inline test/libsolidity/SolidityParser.cpp]: `constructor` should not be reserved, it should be a keyword "in use".

File: test/libsolidity/SolidityEndToEndTest.cpp
Comment (erak) [reply test/libsolidity/SolidityEndToEndTest.cpp]: Oh, this was an unconscious change due to broken code highlighting in my editor. We can revert to ``"abc"``

**Comment by axic** [comment]
Depends on #4206.

**Comment by axic** [comment]
This could be fixed without the external tests:
```
ASSERTION FAILURE:
- file   : SolidityABIJSON.cpp
- line   : 49
- message: Parsing contract failed
```

Also syntax test expectations need to be updated.

File: libsolidity/parsing/Parser.cpp
Comment (axic) [inline libsolidity/parsing/Parser.cpp]: I don't think this can happen since `expectIdentifierToken` shouldn't allow a non-identifier, such as `Token::Constructor`. Is there a test for this which needs this extra condition here?

Comment (erak) [reply libsolidity/parsing/Parser.cpp]: @axic This check should have been removed already. Thanks!

**Comment by erak** [comment]
@axic I updated the broken ABI tests.

**Comment by erak** [comment]
@chriseth The file ``function_named_constructor.sol`` contains a test for 

``contract C { function constructor() { ... }}``

**Comment by chriseth** [comment]
Please rebase to trigger the tests.

**Comment by axic** [comment]
solc-js needs to be updated (thedao code). Also please try to update gnosis first (using their `development` branch), but if it seems to be too much work we can disable it.

Also needs a rebase.

File: Changelog.md
Comment (axic) [inline Changelog.md]: Need to extend the "New keywords" line above with `constructor`.

**Comment by axic** [comment]
Isn't there a documentation section to be updated/extended after this change?

**Comment by chriseth** [comment]
I thought gnosis was not worth keeping because of truffle versioning problems?

**Comment by axic** [comment]
Gnosis has two branches, `master` and `development`, the latter uses the new truffle, that should be close to working.

**Comment by erak** [comment]
@axic Does that mean that we're not going use your Gnosis fork for now? I've been checking out their ``development`` branch, but it also needs an update.

**Comment by erak** [comment]
@axic Regarding documentation: a lot of it was already updated with PR #3635, but I will read through the important sections again.

**Comment by axic** [comment]
> Does that mean that we're not going use your Gnosis fork for now? I've been checking out their development branch, but it also needs an update.

You can push their dev branch to our fork's `solidity-050` branch. And we can take it from there :)

**Comment by bit-shift** [comment]
I created a PR in ``solc-js`` that reflects the new constructor syntax: ethereum/solc-js#225

**Comment by bit-shift** [comment]
- [ ] Switch back ``solc-js`` and Gnosis repositories after ethereum/solc-js#225 and axic/pm-contracts#2 got merged.

**Comment by erak** [comment]
@axic We only have an issue with Truffle left, since the version used by ``npm`` (4.1.11) fails while testing the contracts that are been delivered with it. Do you have any information on how their updating schedule looks like? I would probably create a PR that upadates the contracts in https://github.com/trufflesuite/truffle/tree/develop/packages/truffle-core/lib/testing, but don't let it block this one.

TODO:
- [ ] Switch back ``solc-js`` and Gnosis repositories in ``externalTests.sh`` after ethereum/solc-js#225 and axic/pm-contracts#2 got merged.

**Comment by axic** [comment]
@bit-shift to avoid a lot of rebasing work while this is still pending, we could merge updating our local test suite to use `constructor` in a separate PR right now?

**Comment by erak** [comment]
@axic After 2 unsuccessful attempts to disable the failing Truffle tests, I decided to remove the Gnosis repository from our external tests for now. I will create an issue to remind ourselves to re-add them after the next Truffle release (which will hopefully contain the changes from the PR that I'm going to create within the next days.) I forked and changed the contract that's breaking the test already: https://github.com/bit-shift/truffle/tree/updateConstructor

**Comment by axic** [comment]
@bit-shift can you open a PR on truffle? They should be quick enough to merge.

**Comment by axic** [comment]
Also see my patch in #3534 in the `externalTests.sh` to change truffle. This could be added that way.

**Comment by erak** [comment]
@axic Great! I will apply that workaround.

**Comment by chriseth** [comment]
Needs to be rebased.

**Comment by chriseth** [comment]
Sorry, but this is just gigantic. Could you try to extract updating the external files into a separate PR first? Then we can focus on the semantics changes.

**Comment by erak** [comment]
@chriseth I pulled out all updates to the external compilation tests into a separate PR: #4350, but this one is still huge. I could also extract the tests in ``test/libsolidity`` since it's just the syntax that's been updated in these files and I'd like to keep the actual semantic change and the syntax tests as part of one PR.

**Comment by erak** [comment]
Updated.

**Comment by chriseth** [comment]
Tests are failing.

**Comment by erak** [comment]
We need to merge https://github.com/ethereum/solidity/pull/4354 first. After merging, this PR needs to be updated such that it removes the old syntax tests that expect a warning instead of an error.

**Comment by erak** [comment]
Updated this PR again after #4354 got merged. Should be ready for final review now.

**Comment by erak** [comment]
I missed the docs here, but will add these changes.

**Comment by erak** [comment]
I've split out the required changes to the documentation: https://github.com/ethereum/solidity/pull/4402. After that PR got merged, we can merge this one.

File: libsolidity/analysis/NameAndTypeResolver.cpp
Comment (ekpyron) [inline libsolidity/analysis/NameAndTypeResolver.cpp]: So only something that is a function can be shadowed by something that's not a contract with the same name?
What about ``contract C { uint msg; }``? Currently this warns: ``This declaration shadows a builtin symbol.``. And I think this comes from below in this function (I may be wrong) - does that still work? Is there a test for it?

Comment (erak) [reply libsolidity/analysis/NameAndTypeResolver.cpp]: We have a test for the case you've mentioned: ``452_shadowing_builtins_with_storage_variables.sol``

Comment (erak) [reply libsolidity/analysis/NameAndTypeResolver.cpp]: After we talked about that: I've improved the formatting to make the condition easier to understand 😄

**Review by ekpyron — DISMISSED** [review]
Looks fine to me now.

File: libsolidity/analysis/NameAndTypeResolver.cpp
Comment (chriseth) [inline libsolidity/analysis/NameAndTypeResolver.cpp]: Unrelated to this PR, but why do we have a for loop here?

File: libsolidity/analysis/NameAndTypeResolver.cpp
Comment (chriseth) [inline libsolidity/analysis/NameAndTypeResolver.cpp]: Can you explain the reason for this?

File: libsolidity/analysis/SyntaxChecker.cpp
Comment (chriseth) [inline libsolidity/analysis/SyntaxChecker.cpp]: No space in front of `:`.

File: libsolidity/analysis/NameAndTypeResolver.cpp
Comment (erak) [reply libsolidity/analysis/NameAndTypeResolver.cpp]: @chriseth Without this change, the following code would result in a warning about ``function A()`` shadowing ``contract A``:
```
contract A { constructor() public {} }
contract B is A { function A() public pure returns (uint8) {} }
```
Since functions can now have the same name as a previously declared contract, this should not warn. We also have a test for that: ``overriding_constructor.sol``.

Comment (chriseth) [reply libsolidity/analysis/NameAndTypeResolver.cpp]: I would say the warning is correct, since it does shadow `A`. Note the following does not work because the function `A` shadows the contract `A`:
```
contract A { function f() public {} }
contract B is A {
  function A() public pure returns (uint8) {}
  function g() public {
    A.f();
  }
}
```

**Comment by erak** [comment]
Rebased and squashed some commits. @chriseth I've also addressed your comment about a function shadowing a contract: https://github.com/ethereum/solidity/pull/4215#discussion_r202017629.

File: test/libsolidity/syntaxTests/constructor/constructor.sol
Comment (chriseth) [inline test/libsolidity/syntaxTests/constructor/constructor.sol]: Please move the positive case into its own test.

**Comment by erak** [comment]
@chriseth I reverted merging both, the new and the old constructor syntax, into one file. I also rebased the branch.

**Review by chriseth — APPROVED** [review]

**Pull request merged**
